Add OrderCancellations#cancel_unit + #reimburse_units

philbirt opened the pull request:
* Ability to cancel a unit after it has been marked shipped
* Easy way to reimburse canceled units

We currently have a need to cancel an inventory unit and reimburse it after it has been marked as shipped. This is different than a vanilla return since it is for a case when you have shipped the unit but for any reason you do not want it to be shipped any longer (so it's not being returned). We also have a need  to easily reimburse these canceled units, so I am piggy-backing off the current Reimbursement logic which already does all of the hard stuff for us.

athal7 commented on core/app/models/spree/order_cancellations.rb:
```diff
@@ -42,6 +42,36 @@ def short_ship(inventory_units, whodunnit:nil)
     unit_cancels
   end
 
+  def cancel_unit(inventory_unit, reason: Spree::UnitCancel::DEFAULT_REASON, whodunnit:nil)
```
would it make sense for short_ship to use this method under the hood?

philbirt replied:
Yeah, I definitely considered that when building this. If we can reach a consensus on this change, I'd be happy to refactor `short_ship` to use this and remove the private method its using altogether as well.

jordan-brough replied:
:+1: yanking `short_ship_unit` and re-using this method instead seems like a great idea and a good small step for this PR in making `short_ship` a bit better.

athal7 commented on core/app/models/spree/order_cancellations.rb:
```diff
@@ -42,6 +42,36 @@ def short_ship(inventory_units, whodunnit:nil)
+    unit_cancel
+  end
+
+  def reimburse_units(inventory_units)
```
where does this and the `cancel_unit` method get called?

philbirt replied:
Currently nowhere in this repo, I need it to hook in from an extension.

athal7 replied:
got it

athal7 commented:
seems fine to me, since we are adding public api (and particularly since it isn't called anywhere), maybe some method-level documentation would be good

jordan-brough commented:
Looking good to me. Agreed on adding yardoc on the new method.

philbirt commented:
Refactored this so #short_ship will call #cancel_unit. Unfortunately we are unable to nest OrderMutex locks currently, so this refactoring is dependent on something like https://github.com/solidusio/solidus/pull/507 getting pulled in.

philbirt commented:
Reverted the refactoring for now so that we can spend more time talking about nesting OrderMutexes. Added docs for the public methods in OrderCancellations.

athal7 commented:
ping @jhawthorn @gmacdougall

athal7 commented:
:+1:

gmacdougall commented on core/app/models/spree/order_cancellations.rb:
```diff
@@ -42,6 +46,42 @@ def short_ship(inventory_units, whodunnit:nil)
 
+  # Marks inventory unit canceled. Allows specifying the reason why.
+  #
+  # @return [UnitCancel] the unit that has been canceled
```
Could you add YARD for the parameters as well and mark this as `@api public`?

gmacdougall commented on core/app/models/spree/order_cancellations.rb:
```diff
@@ -42,6 +46,42 @@ def short_ship(inventory_units, whodunnit:nil)
+    Spree::OrderMutex.with_lock!(@order) do
+      return_items = inventory_units.map(&:current_or_new_return_item)
+      reimbursement = Spree::Reimbursement.new(order: @order, return_items: return_items)
+      reimbursement.return_items.map(&:accept!)
```
This is a bit gross. I would like to be able to call `reimbursement.return_all` which does all we need it to do.

gmacdougall commented on core/spec/models/spree/order_cancellations_spec.rb:
```diff
@@ -1,6 +1,64 @@
+    it "creates and performs a reimbursement" do
+      expect { subject }.to change { Spree::Reimbursement.count }.by(1)
+
+      reimbursement = Spree::Reimbursement.last
```
I think this is the `Reumbursement` returned by `subject`.

gmacdougall commented on core/spec/models/spree/order_cancellations_spec.rb:
```diff
@@ -1,6 +1,64 @@
+    end
+
+    it "creates return items for the inventory units and accepts them" do
+      expect { subject }.to change { Spree::ReturnItem.count }.by(1)
```
Is there anything that would cause this to not be true?

gmacdougall commented on core/spec/models/spree/order_cancellations_spec.rb:
```diff
@@ -1,6 +1,64 @@
+    it "creates return items for the inventory units and accepts them" do
+      expect { subject }.to change { Spree::ReturnItem.count }.by(1)
+
+      return_item = Spree::ReturnItem.last
```
I would prefer to get this from the return from `subject` rather than hitting the DB again.

gmacdougall commented on core/spec/models/spree/order_cancellations_spec.rb:
```diff
@@ -1,6 +1,64 @@
+      subject { order.cancellations.cancel_unit(inventory_unit, reason: "some reason") }
+
+      it "sets the reason on the UnitCancel" do
+        expect { subject }.to change { Spree::UnitCancel.count }.by(1)
```
I don't think this expectation is needed. Its been asserted elsewhere.

gmacdougall commented on core/spec/models/spree/order_cancellations_spec.rb:
```diff
@@ -1,6 +1,64 @@
+
+      it "sets the reason on the UnitCancel" do
+        expect { subject }.to change { Spree::UnitCancel.count }.by(1)
+        expect(Spree::UnitCancel.last.reason).to eq("some reason")
```
Can we base this on the return from subject rather than `Spree::UnitCancel.last`.

gmacdougall commented:
I have a few comments that apply to several instances of the same issue. If you could take a look at the related problems as well, I'd appreciate it.

philbirt commented:
@gmacdougall Thanks for the thorough review. I have updated the pull request based on your comments:

* Use subject's return value in place of {ActiveRecord}.last in all of the specs
* Updated to `@api public` where relevant
* Added `@param` to public methods where relevant
* Added a `#return_all!` method to Reimbursement which will accept return items, persist, and perform. I think this is what you were describing in your comment, but let me know if that is not correct. Added api public and a brief description of what it does. I wasn't sure about the return value/raise of this method since it is heavily dependent on the state of the reimbursement and it's return items. Thoughts?

gmacdougall commented on core/app/models/spree/reimbursement.rb:
```diff
@@ -136,6 +136,15 @@ def all_exchanges?
 
+    # Accepts all return items, saves the reimbursement, and performs the reimbursement
+    #
+    # @api public
```
Could you please add a `@return [void]` here?

gmacdougall commented on core/app/models/spree/reimbursement.rb:
```diff
@@ -136,6 +136,15 @@ def all_exchanges?
+    # Accepts all return items, saves the reimbursement, and performs the reimbursement
+    #
+    # @api public
+    def return_all!
```
The ruby convention for methods ending in `!` is that they are a dangerous version of an equivalent method without the `!` (http://dablog.rubypal.com/2007/8/15/bang-methods-or-danger-will-rubyist)

Since there is no `return_all` method that provides this functionality, I would name this `return_all` rather than `return_all!`

philbirt replied:
Interesting. I also thought we would want the bang-named-variant for anything that could reasonably raise exceptions , not only if that method had a non-dangerous version. In this case, return items unable to be accepted or `IncompleteReimbursementError`

gmacdougall commented on core/spec/models/spree/order_cancellations_spec.rb:
```diff
@@ -1,6 +1,58 @@
+      expect { subject }.to change { inventory_unit.state }.to "canceled"
+    end
+
+    context "with a reason" do
```
What happens when no reason is specified?

gmacdougall replied:
I see now that the expectation is listed above on line 12. I might prefer to to put these expectations in pairs:

```Ruby
context 'when a reason is specified' do
  it 'sets the reason to the reason specified`
end

context 'when a reason is not specified' do
  it 'leaves the reason as the default reason'
end
```

gmacdougall commented on core/spec/models/spree/order_cancellations_spec.rb:
```diff
@@ -1,6 +1,58 @@
+      end
+    end
+
+    context "with a who" do
```
Similarly, I would have a context for when whodunnit is not specified here.

gmacdougall commented on core/spec/models/spree/order_cancellations_spec.rb:
```diff
@@ -1,6 +1,58 @@
+    end
+
+    it "creates return items for the inventory units and accepts them" do
+      expect { subject }.to change { Spree::ReturnItem.count }.by(1)
```
I think it's important to test the case where we are returning multiple things.

gmacdougall commented:
Good improvements! I just have a few more suggestions.

philbirt commented:
@gmacdougall Just pushed some code addressing spec comments. I have one question about the `!` method. I'm really not too particular, so I'm happy to change it. I just want to make sure we are consistent about this in the future, as this is something I have seen us do a lot all over the place without a non-`!` variant  https://github.com/solidusio/solidus/pull/502#discussion-diff-45085763

gmacdougall commented:
@jhawthorn  and I discussed the `!` vs. non-`!` methods briefly. It doesn't bother him as much as it does me. The logic around stuff that can be reasonably be expected to throw exceptions having a `!` isn't correct from a Ruby perspective. For example:

```Ruby
[1] pry(main)> [].fetch(1)
IndexError: index 1 outside of array bounds: 0...0
```

@philbirt if you feel strongly that it should have a `!`, then leave it in. Otherwise remove it.

philbirt commented:
@gmacdougall Cool. Personally i've always been in your camp regarding this, so `!` dropped. The contents of the method were pretty `!`-y (with `reimbursement.perform!` not having a `.perform` method) so it seemed like this should have also had `!` in turn.

Good article to keep in my back pocket. Appreciate the thorough review.

gmacdougall commented:
:+1:

athal7 commented:
@jhawthorn the most recent spec failures do *seem* unrelated to me, but i know that generally we have a stable build: https://circleci.com/gh/solidusio/solidus/1365. thoughts?
